add getNull, add compatibility mode for mattn/sqlite3

steffengy opened the pull request:
Early feedback please ;)

This adds the following internally used query parameters:
- _mattn: Enables the compatibility mode when set to "true"
- _loc: Enable time.Time handling and set the locale for it
  This is enabled by setting a locale e.g. _loc=auto for
  the current systems locale.
  This only works when `_mattn` is specified.

cznic commented on sqlite.go:
```diff
@@ -823,41 +928,37 @@ func (s *stmt) bind(stmt uintptr, n int, args []driver.Value) error {
 		}
 	}
-	return nil
+	return err
```
If len(args) > 1 and binding any of the arguments except the last one returns an error the error will be silently ignored.

steffengy replied:
fixed

cznic replied:
Have you pushed the fix? I can only still see the incorrect version of https://github.com/cznic/sqlite/commit/252ae7f8010238f718c1f77f0a5370ce156ee892.

steffengy replied:
https://github.com/cznic/sqlite/commit/252ae7f8010238f718c1f77f0a5370ce156ee892#diff-f6b80b7a52a7492413057b6d5c781efdR942

cznic replied:
Ah, sorry, then it's correct ;-)

Feel free to ignore: A bit tricky to spot however. The previous version had more lines for sure, but it was not so easy to miss the correctness.

cznic commented on sqlite.go:
```diff
@@ -993,16 +1094,45 @@ func (t *tx) exec(stmt string) (err error) {
 	c := &conn{sqlite: s}
 
+	pos := strings.IndexRune(dsn, '?')
+	fmt.Println("DSN: ", dsn)
```
For debugging I use the dbg function (it's defined in all_test.go). It works only while testing but the package does not build otherwise.

cznic commented on sqlite.go:
```diff
@@ -993,16 +1094,45 @@ func (t *tx) exec(stmt string) (err error) {
 	*virtual.Thread
-	pdb uintptr
+	loc           *time.Location
+	compatibility string
```
Please make it a bool.

steffengy replied:
The reasoning behind this was to stay open to allow being compatible to multiple libraries.
Won't probably make sense on a second thought, so I'm changing it.

cznic commented on sqlite.go:
```diff
@@ -363,8 +385,31 @@ func (r *rows) Next(dest []driver.Value) (err error) {
 					return err
 				}
-
+				cty, err := r.columnDeclType(i)
```
This and other related changes have to be executed only in compatibility mode but I can't find the if/switch guard.

steffengy replied:
This one makes actually sense.

Atleast if we want to handle int->boolean conversion by default,
which seems like a straightforward thing to do?

cznic replied:
> Atleast if we want to handle int->boolean conversion by default, which seems like a straightforward thing to do?

Whatever it makes compatible in compatibility mode. No guesswork please in default mode because [sqlite3_column_decltype](https://sqlite.org/c3ref/column_decltype.html) is unreliable:

> If the Nth column of the result set is an expression or subquery, then a NULL pointer is returned.

That means going from, for example, `boolColumn` to `boolColumn or otherBoolColumn` in a query results in suddenly changed type of the value and Scan will fail. Analogically for time/date stuff.

steffengy replied:
Okay, good to know.

cznic commented on sqlite.go:
```diff
@@ -377,8 +422,35 @@ func (r *rows) Next(dest []driver.Value) (err error) {
 					return err
 				}
-
+				cty, err := r.columnDeclType(i)
```
This and other related changes have to be executed only in compatibility mode but I can't find the if/switch guard.

cznic commented on sqlite.go:
```diff
@@ -1312,6 +1442,14 @@ func newDrv() *sqlite { return &sqlite{} }
 //
+//
+// we adds the following query parameters to those used by SQLite:
+//   _mattn=true
```
@mattn Can you please let us know if you don't mind using your handle this way? We can use some other string if you prefer. Thank you.

mattn replied:
Okay. I don't mind. BTW, This project is crazy great. :laughing:

cznic replied:
Thanks a bunch @mattn :wink:

cznic reviewed (changes requested):
Early feedback:

It looks like this PR adds code from mattn's project. If that's the case the proper attribution have to be added, his license terms cited and copy of the license he used included in this repository.

Please add his sqlite3_test.go to this repository and modify it to always turn on the compatibility mode when opening a DB.

steffengy commented:
> It looks like this PR adds code from mattn's project. If that's the case the proper attribution have to be added, his license terms cited and copy of the license he used included in this repository.
Please add his sqlite3_test.go to this repository and modify it to always turn on the compatibility mode when opening a DB.

Well it implements the same logic, reusing some comments.
Since we anyways want to run his tests, we need it anyways for sure.

cznic commented:
Once mattn's tests pass this can go in, I think.

steffengy commented:
> Once mattn's tests pass this can go in, I think.

Working on getting them to run first :)
I have to stub/emulate some functionality in sqlite3 first so it even compiles 
(missing functions, types, hooks)

steffengy committed:
add getNull, add compatibility mode for mattn/sqlite3
This adds the following internally used query parameters:
- _mattn: Enables the compatibility mode when set to "true"
- _loc: Enable time.Time handling and set the locale for it
  This is enabled by setting a locale e.g. _loc=auto for
  the current systems locale.
  This only works when `_mattn` is specified.

steffengy commented:
Had to remove 3 tests (which rely on C-callbacks/registering sqlite functions which is quite a bit to implement):

- TestFunctionDeclaration
- TestDeclTypes
- BenchmarkCustomFunctions

Which leaves us currently at:
```c
--- FAIL: TestOpen (0.05s)
        sqlite3_test.go:97: _txlock error not caught at dbOpen with option: ?_txlock=bogus
--- FAIL: TestForeignKeys (0.00s)
        sqlite3_test.go:153: "PRAGMA foreign_keys;" for "file:C:\\Users\\steff\\AppData\\Local\\Temp\\go-sqlite3-test-548516395?_foreign_keys=1" = false; want true
--- FAIL: TestTransaction (0.03s)
        sqlite3_test.go:697: Unable to query results: <nil>
--- FAIL: TestPreparedStmt (0.26s)
        sqltest.go:294: Query: database is locked (5)
        sqltest.go:294: Query: database is locked (5)
        sqltest.go:294: Query: database is locked (5)
        sqltest.go:294: Query: database is locked (5)
        sqltest.go:298: Insert: database is locked (5)
        sqltest.go:298: Insert: database is locked (5)
        sqltest.go:298: Insert: database is locked (5)
        sqltest.go:298: Insert: database is locked (5)
        sqltest.go:298: Insert: database is locked (5)
BenchmarkExec             5000       3390 req/s
BenchmarkQuery            3000       2231 req/s
BenchmarkParams           3000       2126 req/s
BenchmarkStmt             3000       2317 req/s
BenchmarkRows       panic: sql: Scan error on column index 1: converting driver.Value type <nil> ("<nil>") to a int: invalid syntax

goroutine 19464 [running]:
github.com/mattn/go-sqlite3/sqlite3_test.BenchmarkRows(0xc04209e2c0)
        D:/projects/Go/gopath/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:385 +0x2be
testing.(*B).runN(0xc04209e2c0, 0x1)
        C:/Go/src/testing/benchmark.go:140 +0xb9
testing.(*B).run1.func1(0xc04209e2c0)
        C:/Go/src/testing/benchmark.go:213 +0x61
created by testing.(*B).run1
        C:/Go/src/testing/benchmark.go:214 +0x86
exit status 2
FAIL    github.com/cznic/sqlite 17.759s
```

In comparison the benchmark numbers for mattn's:

```
BenchmarkExec           200000     125905 req/s
BenchmarkQuery          100000      49938 req/s
BenchmarkParams         100000      47103 req/s
BenchmarkStmt           100000      74375 req/s
BenchmarkRows                1          1 req/s
BenchmarkStmtRows         3000       1902 req/s
```

cznic commented:
Nice progress. I think I will review the working version only once we get there. Is that okay for you?

Linux/amd64

```
=== RUN   TestScalar
--- PASS: TestScalar (0.02s)
=== RUN   TestBlob
--- PASS: TestBlob (0.01s)
=== RUN   TestOpen
--- FAIL: TestOpen (0.03s)
	sqlite3_test.go:97: _txlock error not caught at dbOpen with option: ?_txlock=bogus
=== RUN   TestReadonly
--- PASS: TestReadonly (0.01s)
=== RUN   TestForeignKeys
--- FAIL: TestForeignKeys (0.00s)
	sqlite3_test.go:153: "PRAGMA foreign_keys;" for "file:/tmp/go-sqlite3-test-012931647?_foreign_keys=1" = false; want true
=== RUN   TestClose
--- PASS: TestClose (0.01s)
=== RUN   TestInsert
--- PASS: TestInsert (0.01s)
=== RUN   TestUpdate
--- PASS: TestUpdate (0.02s)
=== RUN   TestDelete
--- PASS: TestDelete (0.02s)
=== RUN   TestBooleanRoundtrip
--- PASS: TestBooleanRoundtrip (0.02s)
=== RUN   TestTimestamp
--- PASS: TestTimestamp (0.13s)
=== RUN   TestBoolean
--- PASS: TestBoolean (0.02s)
=== RUN   TestFloat32
--- PASS: TestFloat32 (0.01s)
=== RUN   TestNull
--- PASS: TestNull (0.00s)
=== RUN   TestTransaction
--- FAIL: TestTransaction (0.01s)
	sqlite3_test.go:697: Unable to query results: <nil>
=== RUN   TestWAL
--- PASS: TestWAL (0.02s)
=== RUN   TestTimezoneConversion
--- PASS: TestTimezoneConversion (0.35s)
=== RUN   TestSuite
=== RUN   TestBlobs
--- PASS: TestBlobs (0.01s)
=== RUN   TestManyQueryRow
--- PASS: TestManyQueryRow (5.30s)
=== RUN   TestTxQuery
--- PASS: TestTxQuery (0.01s)
=== RUN   TestPreparedStmt
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [chan receive]:
testing.(*T).Run(0xc42006a0d0, 0x62c860, 0x9, 0x6d37d0, 0xc421b1fd01)
	/home/jnml/go/src/testing/testing.go:698 +0x2f4
testing.runTests.func1(0xc42006a0d0)
	/home/jnml/go/src/testing/testing.go:882 +0x67
testing.tRunner(0xc42006a0d0, 0xc421b1fde0)
	/home/jnml/go/src/testing/testing.go:657 +0x96
testing.runTests(0xc42000a6e0, 0x7cc4a0, 0x1a, 0x1a, 0x0)
	/home/jnml/go/src/testing/testing.go:888 +0x2c1
testing.(*M).Run(0xc421b1ff20, 0xc421b1ff20)
	/home/jnml/go/src/testing/testing.go:822 +0xfc
main.main()
	github.com/cznic/sqlite/_test/_testmain.go:96 +0xf7

goroutine 87 [chan receive]:
testing.(*T).Run(0xc42044d380, 0x62e762, 0x10, 0x6d38a8, 0xc420042c01)
	/home/jnml/go/src/testing/testing.go:698 +0x2f4
testing.runTests.func1(0xc42044d380)
	/home/jnml/go/src/testing/testing.go:882 +0x67
testing.tRunner(0xc42044d380, 0xc420042da8)
	/home/jnml/go/src/testing/testing.go:657 +0x96
testing.runTests(0x6d3888, 0x7cb7a0, 0x4, 0x4, 0xa0)
	/home/jnml/go/src/testing/testing.go:888 +0x2c1
testing.RunTests(0x6d3888, 0x7cb7a0, 0x4, 0x4, 0xc420249ea8)
	/home/jnml/go/src/testing/testing.go:859 +0x56
github.com/mattn/go-sqlite3/sqlite3_test.RunTests(0xc42044d2b0, 0xc42021ca00, 0x0)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:56 +0xed
github.com/cznic/sqlite.TestSuite(0xc42044d2b0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:872 +0x17e
testing.tRunner(0xc42044d2b0, 0x6d37d0)
	/home/jnml/go/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/home/jnml/go/src/testing/testing.go:697 +0x2ca

goroutine 10093 [semacquire]:
sync.runtime_SemacquireMutex(0xc420011f6c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420011f68)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/sqlite.newConn(0xc420011f60, 0xc4202660a0, 0x43, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1165 +0x276
github.com/cznic/sqlite.(*sqlite).Open(0xc420011f60, 0xc4213f81c0, 0x3e, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1483 +0x79
github.com/cznic/sqlite.(*CompatHelper).Open(0xc420011f60, 0xc421b4a7c0, 0x32, 0x0, 0x7f9128265fe8, 0x7f9128265fb0, 0x7f9128265f50)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:37 +0xc8
database/sql.(*DB).conn(0xc42021ca00, 0x7b3320, 0xc420010e78, 0xc420027501, 0x0, 0x0, 0x3)
	/home/jnml/go/src/database/sql/sql.go:965 +0x146
database/sql.(*Stmt).connStmt(0xc4213df900, 0x7b3320, 0xc420010e78, 0x50, 0xc420014140, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1903 +0x1e6
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1941 +0xe5
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc420462fb0)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 35 [chan receive]:
database/sql.(*DB).connectionOpener(0xc420090f00)
	/home/jnml/go/src/database/sql/sql.go:837 +0x4a
created by database/sql.Open
	/home/jnml/go/src/database/sql/sql.go:582 +0x212

goroutine 36 [chan receive]:
database/sql.(*DB).connectionOpener(0xc4200910e0)
	/home/jnml/go/src/database/sql/sql.go:837 +0x4a
created by database/sql.Open
	/home/jnml/go/src/database/sql/sql.go:582 +0x212

goroutine 63 [chan receive]:
database/sql.(*Rows).awaitDone(0xc4201f3c00, 0x7b32e0, 0xc420216040)
	/home/jnml/go/src/database/sql/sql.go:2121 +0x58
created by database/sql.(*Rows).initContextClose
	/home/jnml/go/src/database/sql/sql.go:2116 +0x92

goroutine 66 [chan receive]:
database/sql.(*Rows).awaitDone(0xc42021a080, 0x7b32e0, 0xc4202168c0)
	/home/jnml/go/src/database/sql/sql.go:2121 +0x58
created by database/sql.(*Rows).initContextClose
	/home/jnml/go/src/database/sql/sql.go:2116 +0x92

goroutine 69 [chan receive]:
database/sql.(*Rows).awaitDone(0xc42021a200, 0x7b32e0, 0xc420216bc0)
	/home/jnml/go/src/database/sql/sql.go:2121 +0x58
created by database/sql.(*Rows).initContextClose
	/home/jnml/go/src/database/sql/sql.go:2116 +0x92

goroutine 10088 [semacquire]:
sync.runtime_SemacquireMutex(0xc420013874)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420013870)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/virtual.(*cpu).pthreadMutexLock(0xc421280000)
	/home/jnml/src/github.com/cznic/virtual/pthread.go:114 +0x1f4
github.com/cznic/virtual.(*cpu).(github.com/cznic/virtual.pthreadMutexLock)-fm()
	/home/jnml/src/github.com/cznic/virtual/cpu.go:1718 +0x2a
github.com/cznic/virtual.(*cpu).builtin(0xc421280000, 0xc421b1b118)
	/home/jnml/src/github.com/cznic/virtual/cpu.go:104 +0x25
github.com/cznic/virtual.(*cpu).run(0xc421280000, 0x3802e, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/virtual/cpu.go:1718 +0x63fa
github.com/cznic/virtual.(*Thread).FFI(0xc421280000, 0x3802e, 0xc421b1b7e0, 0x1, 0x1, 0xc421b1b880, 0x4, 0x4, 0x5e4100, 0xc4204630a8, ...)
	/home/jnml/src/github.com/cznic/virtual/thread.go:158 +0x217
github.com/cznic/virtual.(*Thread).FFI1(0xc421280000, 0x3802e, 0x7b1aa0, 0xc420463078, 0xc421b1b880, 0x4, 0x4, 0x0, 0x7f9128186000, 0x40000)
	/home/jnml/src/github.com/cznic/virtual/thread.go:110 +0xab
github.com/cznic/sqlite.(*conn).openV2(0xc42125e000, 0xc420266000, 0x43, 0x10046, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1366 +0x289
github.com/cznic/sqlite.newConn(0xc420011f60, 0xc420266000, 0x43, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1192 +0x362
github.com/cznic/sqlite.(*sqlite).Open(0xc420011f60, 0xc4213f8000, 0x3e, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1483 +0x79
github.com/cznic/sqlite.(*CompatHelper).Open(0xc420011f60, 0xc421b4a7c0, 0x32, 0xc42001fe00, 0xc4213f78e0, 0x7b2e60, 0xc4213f9600)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:37 +0xc8
database/sql.(*DB).conn(0xc42021ca00, 0x7b3320, 0xc420010e78, 0x5e8ad1bfa2a7dd01, 0xa428f0c378a81837, 0x1, 0x3)
	/home/jnml/go/src/database/sql/sql.go:965 +0x146
database/sql.(*Stmt).connStmt(0xc4213df900, 0x7b3320, 0xc420010e78, 0x50, 0x50, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1903 +0x1e6
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1941 +0xe5
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc420462f60)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 73 [chan receive]:
database/sql.(*Rows).awaitDone(0xc42021a780, 0x7b32e0, 0xc420217380)
	/home/jnml/go/src/database/sql/sql.go:2121 +0x58
created by database/sql.(*Rows).initContextClose
	/home/jnml/go/src/database/sql/sql.go:2116 +0x92

goroutine 10090 [semacquire]:
sync.runtime_SemacquireMutex(0xc420011f6c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420011f68)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/sqlite.newConn(0xc420011f60, 0xc421b28050, 0x43, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1165 +0x276
github.com/cznic/sqlite.(*sqlite).Open(0xc420011f60, 0xc421260140, 0x3e, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1483 +0x79
github.com/cznic/sqlite.(*CompatHelper).Open(0xc420011f60, 0xc421b4a7c0, 0x32, 0x0, 0x7f9128265ff0, 0x7f9128265fd0, 0x7f9128265fb0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:37 +0xc8
database/sql.(*DB).conn(0xc42021ca00, 0x7b3320, 0xc420010e78, 0xc42020e501, 0xf10f90a838b72ffd, 0x659a743dfb04becb, 0x3)
	/home/jnml/go/src/database/sql/sql.go:965 +0x146
database/sql.(*Stmt).connStmt(0xc4213df900, 0x7b3320, 0xc420010e78, 0xed70c1b44dfcab2d, 0x85a2adc5c38ad0d4, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1903 +0x1e6
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1941 +0xe5
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc42125c008)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 88 [chan receive]:
database/sql.(*DB).connectionOpener(0xc42021ca00)
	/home/jnml/go/src/database/sql/sql.go:837 +0x4a
created by database/sql.Open
	/home/jnml/go/src/database/sql/sql.go:582 +0x212

goroutine 10086 [semacquire]:
sync.runtime_Semacquire(0xc4213ec62c)
	/home/jnml/go/src/runtime/sema.go:47 +0x34
sync.(*WaitGroup).Wait(0xc4213ec620)
	/home/jnml/go/src/sync/waitgroup.go:131 +0x7a
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt(0xc42006a270)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:304 +0x37c
testing.tRunner(0xc42006a270, 0x6d38a8)
	/home/jnml/go/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/home/jnml/go/src/testing/testing.go:697 +0x2ca

goroutine 10096 [semacquire]:
sync.runtime_SemacquireMutex(0xc42001387c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420013878)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/virtual.(*cpu).pthreadMutexUnlock(0xc4200a3ea0)
	/home/jnml/src/github.com/cznic/virtual/pthread.go:171 +0x81
github.com/cznic/virtual.(*cpu).(github.com/cznic/virtual.pthreadMutexUnlock)-fm()
	/home/jnml/src/github.com/cznic/virtual/cpu.go:1720 +0x2a
github.com/cznic/virtual.(*cpu).builtin(0xc4200a3ea0, 0xc42127f450)
	/home/jnml/src/github.com/cznic/virtual/cpu.go:104 +0x25
github.com/cznic/virtual.(*cpu).run(0xc4200a3ea0, 0x15012, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/virtual/cpu.go:1720 +0x63af
github.com/cznic/virtual.(*Thread).FFI(0xc4200a3ea0, 0x15012, 0xc42127fb28, 0x1, 0x1, 0xc420069ba8, 0x1, 0x1, 0x5e4100, 0xc4213ec6b0, ...)
	/home/jnml/src/github.com/cznic/virtual/thread.go:158 +0x217
github.com/cznic/virtual.(*Thread).FFI1(0xc4200a3ea0, 0x15012, 0x7b1aa0, 0xc4213ec6a8, 0xc420069ba8, 0x1, 0x1, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/virtual/thread.go:110 +0xab
github.com/cznic/sqlite.(*stmt).step(0xc4213eb440, 0x7f912a49bab0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1020 +0xe4
github.com/cznic/sqlite.(*stmt).Query(0xc4213eb440, 0x7ea900, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:791 +0x1ae
database/sql.ctxDriverStmtQuery(0x7b3320, 0xc420010e78, 0x7b33a0, 0xc4213eb440, 0x7ea900, 0x0, 0x0, 0x0, 0x0, 0x7f912a998e38, ...)
	/home/jnml/go/src/database/sql/ctxutil.go:95 +0x18c
database/sql.rowsiFromStatement(0x7b3320, 0xc420010e78, 0xc4213eb540, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:2002 +0x1eb
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1949 +0x241
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc4213ec630)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 10092 [semacquire]:
sync.runtime_SemacquireMutex(0xc420011f6c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420011f68)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/sqlite.newConn(0xc420011f60, 0xc4202660f0, 0x43, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1165 +0x276
github.com/cznic/sqlite.(*sqlite).Open(0xc420011f60, 0xc4213f82c0, 0x3e, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1483 +0x79
github.com/cznic/sqlite.(*CompatHelper).Open(0xc420011f60, 0xc421b4a7c0, 0x32, 0x7f912a593e00, 0x22, 0x0, 0x7f912a594890)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:37 +0xc8
database/sql.(*DB).conn(0xc42021ca00, 0x7b3320, 0xc420010e78, 0xc421b93501, 0x0, 0x0, 0x3)
	/home/jnml/go/src/database/sql/sql.go:965 +0x146
database/sql.(*Stmt).connStmt(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x7cf5c0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1903 +0x1e6
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1941 +0xe5
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc420462fb8)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 10094 [semacquire]:
sync.runtime_SemacquireMutex(0xc420011f6c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420011f68)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/sqlite.newConn(0xc420011f60, 0xc421b280a0, 0x43, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1165 +0x276
github.com/cznic/sqlite.(*sqlite).Open(0xc420011f60, 0xc421260240, 0x3e, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1483 +0x79
github.com/cznic/sqlite.(*CompatHelper).Open(0xc420011f60, 0xc421b4a7c0, 0x32, 0x0, 0x7f9128265ff0, 0x7f9128265fd0, 0x7f9128265fb0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:37 +0xc8
database/sql.(*DB).conn(0xc42021ca00, 0x7b3320, 0xc420010e78, 0xc421b79501, 0x0, 0x0, 0x3)
	/home/jnml/go/src/database/sql/sql.go:965 +0x146
database/sql.(*Stmt).connStmt(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1903 +0x1e6
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1941 +0xe5
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc42125c050)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 10087 [semacquire]:
sync.runtime_SemacquireMutex(0xc420011f6c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420011f68)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/sqlite.newConn(0xc420011f60, 0xc420266050, 0x43, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1165 +0x276
github.com/cznic/sqlite.(*sqlite).Open(0xc420011f60, 0xc4213f80c0, 0x3e, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1483 +0x79
github.com/cznic/sqlite.(*CompatHelper).Open(0xc420011f60, 0xc421b4a7c0, 0x32, 0x0, 0x7f9128265fe8, 0x7f9128265fb0, 0x7f9128265f50)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:37 +0xc8
database/sql.(*DB).conn(0xc42021ca00, 0x7b3320, 0xc420010e78, 0xc421db2501, 0x7f9128264788, 0x7f9128264730, 0x3)
	/home/jnml/go/src/database/sql/sql.go:965 +0x146
database/sql.(*Stmt).connStmt(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1903 +0x1e6
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1941 +0xe5
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc420462f68)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 10095 [semacquire]:
sync.runtime_SemacquireMutex(0xc420011f6c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420011f68)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/sqlite.newConn(0xc420011f60, 0xc420266190, 0x43, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1165 +0x276
github.com/cznic/sqlite.(*sqlite).Open(0xc420011f60, 0xc4213f83c0, 0x3e, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1483 +0x79
github.com/cznic/sqlite.(*CompatHelper).Open(0xc420011f60, 0xc421b4a7c0, 0x32, 0x0, 0x0, 0xc421265b00, 0x100000000)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:37 +0xc8
database/sql.(*DB).conn(0xc42021ca00, 0x7b3320, 0xc420010e78, 0xc421bb8501, 0x0, 0x7f9128265fe8, 0x3)
	/home/jnml/go/src/database/sql/sql.go:965 +0x146
database/sql.(*Stmt).connStmt(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x7cf5c0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1903 +0x1e6
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1941 +0xe5
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc420463000)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 10091 [semacquire]:
sync.runtime_SemacquireMutex(0xc42001387c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420013878)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/virtual.(*cpu).pthreadMutexLock(0xc4200a2000)
	/home/jnml/src/github.com/cznic/virtual/pthread.go:109 +0x73
github.com/cznic/virtual.(*cpu).(github.com/cznic/virtual.pthreadMutexLock)-fm()
	/home/jnml/src/github.com/cznic/virtual/cpu.go:1718 +0x2a
github.com/cznic/virtual.(*cpu).builtin(0xc4200a2000, 0xc42041f400)
	/home/jnml/src/github.com/cznic/virtual/cpu.go:104 +0x25
github.com/cznic/virtual.(*cpu).run(0xc4200a2000, 0x14ffb, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/virtual/cpu.go:1718 +0x63fa
github.com/cznic/virtual.(*Thread).FFI(0xc4200a2000, 0x14ffb, 0xc42041fac8, 0x1, 0x1, 0xc42041fb68, 0x5, 0x5, 0x5e4100, 0xc420435128, ...)
	/home/jnml/src/github.com/cznic/virtual/thread.go:158 +0x217
github.com/cznic/virtual.(*Thread).FFI1(0xc4200a2000, 0x14ffb, 0x7b1aa0, 0xc420435100, 0xc42041fb68, 0x5, 0x5, 0x0, 0xf40000c420056000, 0xc42041fb80)
	/home/jnml/src/github.com/cznic/virtual/thread.go:110 +0xab
github.com/cznic/sqlite.(*stmt).prepareV2(0xc4213ee240, 0x7f912a5947d0, 0x7ea900, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1037 +0x20c
github.com/cznic/sqlite.(*stmt).Query(0xc4213ee240, 0x7ea900, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:770 +0xd2
database/sql.ctxDriverStmtQuery(0x7b3320, 0xc420010e78, 0x7b33a0, 0xc4213ee240, 0x7ea900, 0x0, 0x0, 0x0, 0x0, 0xc4213ee340, ...)
	/home/jnml/go/src/database/sql/ctxutil.go:95 +0x18c
database/sql.rowsiFromStatement(0x7b3320, 0xc420010e78, 0xc4213ee340, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:2002 +0x1eb
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1949 +0x241
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc420434f50)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357

goroutine 10089 [semacquire]:
sync.runtime_SemacquireMutex(0xc420011f6c)
	/home/jnml/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc420011f68)
	/home/jnml/go/src/sync/mutex.go:87 +0x9d
github.com/cznic/sqlite.newConn(0xc420011f60, 0xc421b28000, 0x43, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1165 +0x276
github.com/cznic/sqlite.(*sqlite).Open(0xc420011f60, 0xc421260040, 0x3e, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/src/github.com/cznic/sqlite/sqlite.go:1483 +0x79
github.com/cznic/sqlite.(*CompatHelper).Open(0xc420011f60, 0xc421b4a7c0, 0x32, 0x0, 0x7f9128265fe8, 0x1, 0x7f912a49b600)
	/home/jnml/src/github.com/cznic/sqlite/sqlite3_test.go:37 +0xc8
database/sql.(*DB).conn(0xc42021ca00, 0x7b3320, 0xc420010e78, 0xc421bc5d01, 0x0, 0x7f9128265fe8, 0x3)
	/home/jnml/go/src/database/sql/sql.go:965 +0x146
database/sql.(*Stmt).connStmt(0xc4213df900, 0x7b3320, 0xc420010e78, 0x7f9128265f90, 0x200, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1903 +0x1e6
database/sql.(*Stmt).QueryContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jnml/go/src/database/sql/sql.go:1941 +0xe5
database/sql.(*Stmt).QueryRowContext(0xc4213df900, 0x7b3320, 0xc420010e78, 0x0, 0x0, 0x0, 0xc42125c000)
	/home/jnml/go/src/database/sql/sql.go:2021 +0x70
database/sql.(*Stmt).QueryRow(0xc4213df900, 0x0, 0x0, 0x0, 0x6d3ee0)
	/home/jnml/go/src/database/sql/sql.go:2040 +0x61
github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt.func1(0xc4213ec620, 0xc4213df900, 0xc42006a270, 0xc4213df980)
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:293 +0xba
created by github.com/mattn/go-sqlite3/sqlite3_test.TestPreparedStmt
	/home/jnml/src/github.com/mattn/go-sqlite3/sqlite3_test/sqltest.go:302 +0x357
exit status 2
FAIL	github.com/cznic/sqlite	6.136s
```

Seems like there's some work ahead for both of us.

----

Not directly related but right now in my mind:

- I noticed some .go files in our working set of repositories (ccir and virtual IIRC) are not gofmted. Also some .h files in ccir lack "standard" formatting (see the indent command in the Makefile). No priority, let's fix this during/after next merge to respective master branches.

- I have finished optimizing switch statements, which according to perflog contribute a lot to the number of VM instructions executed. Once I catch up rebasing to latest master I will push to 'switch' branches in ccir and virtual. I would like you to review the changes and test everything with Windows. sqlite will work again on Windows only after running 'go generate' (use some temp branch for that). The break comes from adding two new switch operations to IR and the respective VM instructions. I don't expect more breaking changes in IR or virtual, but I want to remind us that the current binaries in sqlite have to be regenerated and pushed if/when the switch branches go in.

> In comparison the benchmark numbers for mattn's:

The VM is very slow compared to native C code, but honestly I would have expected it to be even slower. The switch branch improves performance of the previously existing benchmarks in cznic/sqlite and I still have another issue I plan to address where I hope for a comparable improvement. Maybe I can still do that later today.

steffengy commented:
> Nice progress. I think I will review the working version only once we get there. Is that okay for you?

Yeah, just posting the current status so anyone who's interested in it can stay updated.

> The VM is very slow compared to native C code, but honestly I would have expected it to be even slower. 

Yeah me to. Only added the native values since I saw that the benchmarks are anyways included in the output I need to post. And posting them without any relation seemed useless. (+ was interested in how bad it actually is, seems to be 20x-80x)

steffengy commented:
Just thinking about this since for this we have to anyways reuse loads of mattn's code,
I'm thinking about the following solution:

- Keep this as trivial implementation (that might be helpful for initial porting/whatever), 
eventually drop it (on the very longterm)
- "Fork" mattn/sqlite3
remove CGO references and 
introduce a new subpackage and import it as `C` e.g.
```go
import C "github.com/fork/c"
```
In this new subpackage the VM specific implementation would rest 
(defining the necessary e.g. `sqlite3_version` functions, which are called from mattn - code generation could make sense...) and the code would mostly be the same, just use `cznic/virtual` insteadof CGO as a "C-backend".

==> 100% compatibility, easy to update

This is just an idea, also not sure how feasible this actually is (introducing the C package works that way).

What do you think?

cznic commented:
> What do you think?

Today evening I finally started to add some real code to [ccgo](https://github.com/cznic/ccgo). Just moving tests from ccir for now, but it's at least something. I hope to have a more informed answer to your question tomorrow evening/night. Let's see how far I can get with ccgo, then we can guestimate what ROI we can reasonably expect from additional VM approach effort.

cznic commented:
(No content there yet.)

steffengy commented:
@cznic
Great!
It was just a thought, have to think of some solution that'd also work with `ccgo`.

Which it might would, depending on what code CCGO generates, but we'll see that
when it's time, this isn't urgent by any means - just had the thought and wanted to document
it

Will you be doing the actual code generation part using `go/ast` and `go/printer` ?

cznic commented:
> Will you be doing the actual code generation part using go/ast and go/printer ?

fmt.Fprintf ;-)

go/ast will be used not for generating code, but for making it better *after* it's generated. Those stages do need to walk the AST and, for example, turn certain int32 expression evaluations to boolean Go expressions, etc. However, to preserve the original C semantics that's not really necessary. The initial version will thus probably never go through the AST form at all. At least not directly, only indirectly by using go/format before persisting the resulting Go code.

Readability/maintainability of Go code that ccgo will produce *is not* a goal. I have a pretty good idea how it will look like and *it will be absolutely horrible*. The only reason I consider making the code look better in certain places is because I expect the improved Go code can lead to better machine code generated at the very end.

For now, I don't even plan to generate any Go struct types. Also *all* pointers will be uintpr. For ccgo, Go is "just" a backend :thinking:

steffengy commented:
> For now, I don't even plan to generate any Go struct types.

Go struct types would cause issues with unions etc. anyways.

cznic commented:
@steffengy  Almost no progress with ccgo, as you've probably noticed. Less free time than I expected to have :-(

steffengy commented:
@cznic
No worries, little progress is always better than no progress :)

cznic commented:
@steffengy The baby has made its first six steps: https://github.com/cznic/ccgo/commit/de5a2f1c49e0ab9c16b8c36bed355b2f90bb91e7 :unicorn:

steffengy commented:
@cznic 
Nice! Let me know when testing on windows makes sense.

cznic commented:
For sure.

cznic commented:
@steffengy Since https://github.com/cznic/irgo/commit/fcda165b801fced127996ae3b55516923e3b477a about half of tests pass* (TestTCC 55/55, TestGCCExec 737/1250, TestSQLite and others do not work yet but it's getting close). It is possibly now a good time to start thinking about filling the Windows missing pieces in [crt](https://github.com/cznic/crt) in parallel. WDYT?

*: To actually run the tests you need to test the cznic/ccgo package.

steffengy commented:
@cznic
windows-specific stuff really only is syscalls right?

How do you plan to handle those for linux?
Similar to `virtual` (wrapper func with calls the `syscall` package, I guess this) 
or generating the calls to `syscall.` during compilation?

I'll start thinking about it, should be only a bit work.

cznic commented:
> windows-specific stuff really only is syscalls right?

Mostly. It's just copy/pasting Window specific parts from cznic/virtual to cznic/crt with minor changes.

Big catch: TLS is not yet implemented and later every function will additionally get a first parameter of type *tls injected. There the usual thread ID and errno will live plus more.

> How do you plan to handle those for linux?

Same as above for windows. Few are already in place, might give you some inspiration.

> Similar to virtual (wrapper func with calls the syscall package, I guess this)
or generating the calls to syscall. during compilation?

Linux will be coded manually, I believe. In virtual you've sucessfully generated the Windows stuff, you can perhaps adjust that generator and reuse it for cznic/crt.

> I'll start thinking about it, should be only a bit work.

Everything I think is a bit of work usually takes a day at least ;-)

steffengy commented:
@cznic
Any notes on getting it work on windows?

I looked into it for a bit but it seems that you "pattern-match" structs to determine types.
(e.g. a specific struct signature equals the `FILE` type)
This is problematic for windows since `FILE` essentially is only an integer.

(Which means duplicating all file methods to take `unsafe.Pointer` since that's what's type-matched,
or somehow changing the declaration of `FILE` to point to some kind of a dummy struct with a unique signature that's then someow matched, which both seems really ugly.)

cznic commented:
@steffengy Glad to hear from you!

> Any notes on getting it work on windows?

Nothing too particular. I think that when you try to make ccgo tests pass, the build and other failures will just "ask" for missing features in crt. Then "all" what's needed is to fill the gaps ;-)

> I looked into it for a bit but it seems that you "pattern-match" structs to determine types.
(e.g. a specific struct signature equals the FILE type)
This is problematic for windows since FILE essentially is only an integer.

No worries, the pattern is a real IR type specifier and is used only for structs/unions. If the Windows handle is an integer it should not be a problem, I believe.

I suggest to create a `windows` branch in {crt,irgo,ccgo,sqlite} and let me see there your attempts and the fails as soon as they happen. I will then try to fix and remove any unearthed roadblocks ASAP.

steffengy commented:
Well not really doing any modifications (except commething `	ir.TypeID(dict.SID(crt.TFILE)):                "crt.XFILE",` in `ccgo` for windows)
 and defining `type XFILE unsafe.Pointer` gives e.g.:

```go
main.go:97: cannot use _f (type *unsafe.Pointer) as type *crt.XFILE in argument to crt.Xfclose
```

cznic commented:
I think it's necessary to move stdio.go to stdio_linux.go and copy it to stdio_windows.go where it will be adapted for Windows (or use other names and proper build tags). Export the TFILE constant as an empty string, that should work, I hope.

BTW, I see stdio32.go is missing the `+build !windows` line, please add it in the same way it can be seen in `stdio64.go`.

Windows versions then can be in stdio32_windows and stdio64_windows.go, for example.

steffengy commented:
Do I need TFILE/XFILE on windows for something?
I think just using `unsafe.Pointer` would make sense buf for some reason
it seems to expect `*unsafe.Pointer`

cznic commented:
> Do I need TFILE/XFILE on windows for something?

I don't think so. But setting TFILE to an ampty string should have disabled using XFILE in the first place.

> I think just using unsafe.Pointer would make sense buf for some reason
it seems to expect *unsafe.Pointer

That's surprising and I don't know why it happens. It should use the same ccir/libc headers as with the VM.

Wait, that really is the [signature](https://github.com/cznic/ccir/blob/5ffcdc88b8dc713e98ad8c58e323e1a430027720/libc/stdio_windows_amd64.h#L75): `int fclose (FILE *_File );`.

?

steffengy commented:
Oh nevermind, `typedef void * FILE;`  explains a lot.
Shouldn't make any issues though, just something to be aware of.

steffengy commented:
Just to be sure thats not an issue with just the WIN-version:

Is it expected that those CCGO test runs take really long (>10 minutes)?
(I guess it spawns new processes so yes?)

cznic commented:
> Is it expected that those CCGO test runs take really long (>10 minutes)?

Yes it is (on slower systems). The actual tests execute via `go run` on the produced Go-from-C code.

Use the -timeout 24h or similar option. You may add the -trc option to see the progress.

steffengy commented:
> Yes it is (on slower systems).

Not really using any system I'd call remotely slow.
I guess on windows the bloated headers (which are very big in comparison to linux, MBs vs KBs)
only make this worse.

cznic commented:
> Not really using any system I'd call remotely slow.

Well, I was not thinking about the HW ;-)

> I guess on windows the bloated headers (which are very big in comparison to linux, MBs vs KBs)
only make this worse.

Every header is parsed once and only once.

PS: On Linux the tests finish in about 6 minutes, IIRC.

steffengy commented:
GCC 1136/1138

   .\main.go:107: cannot use uint32(u64(31)) (type uint32) as type uint64 in argument to crt.Xmemset

It it possible that `long long` somehow isn't interpreted as uint64?

cznic commented:
> It it possible that long long somehow isn't interpreted as uint64?

I don't think that should be the case, so there's a bug somewhere.

Otherwise congrats on getting so far so quickly!

steffengy commented:
Seems to be the case:
`size_t` is typedef'd to `unsigned long long` for amd64/windows.

getting similar errors for C4:
```
 .\main.go:884: cannot use uint32(*_sp) (type uint32) as type uint64 in argument to crt.Xread
```

cznic commented:
> Seems to be the case:
size_t is typedef'd to unsigned long long for amd64/windows.

That sounds okay to me. Or is it the same problem as with `long` being only 32 bits on 64 bit Windows?

I cannot even find the definition:

```bash
nml@r550:~/src/github.com/cznic/ccir/libc$ grep -nr 'typedef.* size_t;' *
pthread_linux_amd64.h:145:typedef long unsigned int size_t;
pthread_linux_386.h:145:typedef unsigned int size_t;
stddef_linux_amd64.h:32:typedef long unsigned int size_t;
stddef_linux_386.h:32:typedef unsigned int size_t;
stdio_linux_amd64.h:81:typedef long unsigned int size_t;
stdio_linux_386.h:81:typedef unsigned int size_t;
stdlib_linux_amd64.h:51:typedef long unsigned int size_t;
stdlib_linux_386.h:51:typedef unsigned int size_t;
string_linux_amd64.h:51:typedef long unsigned int size_t;
string_linux_386.h:51:typedef unsigned int size_t;
strings_linux_amd64.h:51:typedef long unsigned int size_t;
strings_linux_386.h:51:typedef unsigned int size_t;
sys/mman_linux_386.h:85:typedef unsigned int size_t;
sys/mman_linux_amd64.h:85:typedef long unsigned int size_t;
time_linux_amd64.h:51:typedef long unsigned int size_t;
time_linux_386.h:51:typedef unsigned int size_t;
unistd_linux_amd64.h:201:typedef long unsigned int size_t;
unistd_linux_386.h:201:typedef unsigned int size_t;
jnml@r550:~/src/github.com/cznic/ccir/libc$ 
```

steffengy commented:
There's probably a space (before the semicolon) that's trolling you:

```
stdlib_windows_amd64.h
7:typedef unsigned long long size_t ;
```

cznic commented:
> GCC 1136/1138
>
> .\main.go:107: cannot use uint32(u64(31)) (type uint32) as type uint64 in argument to crt.Xmemset

Please paste the full output for the failing case. Use -run GCC -re filename.c -v. I want to look at it if I may get some idea.

steffengy commented:
```go
all_test.go:387: ..\cc\testdata\gcc-6.3.0\gcc\testsuite\gcc.c-torture\execute\pr30778.c: FAIL
                go run: exit status 0, err exit status 2
                # ccgo.New
                // Code generated by ccgo DO NOT EDIT.

                package main

                import (
                        "fmt"
                        "math"
                        "os"
                        "path"
                        "runtime"
                        "unsafe"

                        "github.com/cznic/crt"
                        "github.com/edsrzf/mmap-go"
                )

                const (
                        heapReserve = 1 << 20
                        heapSize    = 32 << 20
                        uintptrSize = 1 << (^uintptr(0)>>32&1 + ^uintptr(0)>>16&1 + ^uintptr(0)>>8&1 + 3) / 8
                )

                var (
                        args [1 << 16]byte
                        inf  = math.Inf(1)
                )

                func ftrace(s string, args ...interface{}) {
                        _, fn, fl, _ := runtime.Caller(1)
                        fmt.Fprintf(os.Stderr, "# %s:%d: %v\n", path.Base(fn), fl, fmt.Sprintf(s, args...))
                        os.Stderr.Sync()
                }

                func main() {
                        heap, err := mmap.MapRegion(nil, heapSize+heapReserve, mmap.RDWR, mmap.ANON, 0)
                        if err != nil {
                                panic(err)
                        }

                        crt.RegisterHeap(unsafe.Pointer(&heap[0]), heapSize+heapReserve)
                        os.Args[0] = "./test"
                        var a []int
                        i := 0
                        for _, v := range os.Args {
                                v += "\x00"
                                a = append(a, i)
                                copy(args[i:], v)
                                i = (i + len(v) + 16) &^ 15
                        }
                        argv := (**int8)(unsafe.Pointer(&args[i]))
                        for _, v := range a {
                                *(*uintptr)(unsafe.Pointer(&args[i])) = uintptr(unsafe.Pointer(&args[v]))
                                i += uintptrSize
                        }
                        X_start(crt.NewTLS(), int32(len(os.Args)), argv)
                }

                func X_start(tls *crt.TLS, _argc int32, _argv **int8) { // crt0.c:12:1
                        crt.X__register_stdfiles(tls, Xstdin, Xstdout, Xstderr)
                        crt.X__builtin_exit(tls, Xmain(tls, _argc, _argv))
                        return

                }

                var Xstdin unsafe.Pointer // crt0.c:10:6

                func init() {
                        Xstdin = unsafe.Pointer(&X__stdfiles)
                }

                var X__stdfiles [3]unsafe.Pointer // crt0.c:9:15

                var Xstdout unsafe.Pointer // crt0.c:10:31

                func init() {
                        Xstdout = (unsafe.Pointer)(uintptr(unsafe.Pointer(&X__stdfiles)) + 8)
                }

                var Xstderr unsafe.Pointer // crt0.c:10:57

                func init() {
                        Xstderr = (unsafe.Pointer)(uintptr(unsafe.Pointer(&X__stdfiles)) + 16)
                }

                func Xmain(tls *crt.TLS, _ int32, _ **int8) (r0 int32) { // pr30778.c:24:1
                        var _r T1181033 // pr30778.c:26:19
                        _ = _r
                        r0 = i32(0)
                        _reg_stat = &_r
                        bug20530(_reg_stat)
                        *(*int32)(unsafe.Pointer(uintptr((unsafe.Pointer)(&_r)) + uintptr(32 /* X8 */))) = i32(-1)
                        Xinit_reg_last(tls)
                        if (_r.X8) == i32(-1) {
                                goto _0
                        }

                        crt.Xabort(tls)
                _0:
                        r0 = i32(0)
                        return

                }

                var _reg_stat *T1181033 // pr30778.c:16:24

                func Xinit_reg_last(tls *crt.TLS) { // pr30778.c:18:1
                        crt.Xmemset(tls, (unsafe.Pointer)(_reg_stat), i32(0), uint32(u64(31)))
                        return

                }

                func bool2int(b bool) int32 {
                        if b {
                                return 1
                        }
                        return 0
                }
                func bug20530(interface{}) {} //TODO remove when https://github.com/golang/go/issues/20530 is fixed.
                func i16(n int16) int16    { return n }
                func i32(n int32) int32    { return n }
                func i64(n int64) int64    { return n }
                func i8(n int8) int8       { return n }
                func init()                { nzf32 *= -1; nzf64 *= -1 }
                func u16(n uint16) uint16  { return n }
                func u32(n uint32) uint32  { return n }
                func u64(n uint64) uint64  { return n }
                func u8(n byte) byte       { return n }

                var nzf32 float32 // -0.0
                var nzf64 float64 // -0.0

                type T1181033 struct {
                        X0 unsafe.Pointer
                        X1 unsafe.Pointer
                        X2 unsafe.Pointer
                        X3 int32
                        X4 int8
                        X5 int8
                        X6 int8
                        X7 int8
                        X8 int32
                } // struct{*struct{},*struct{},*struct{},int32,int8,int8,int8,int8,int32}
                stderr:
                # command-line-arguments
                .\main.go:107: cannot use uint32(u64(31)) (type uint32) as type uint64 in argument to crt.Xmemset

```

cznic commented:
in ccir/libc

`windows_windows_amd64.h:33783:#define FIELD_OFFSET(Type, Field) ( ( LONG ) __builtin_offsetof ( Type , Field ) )`

The cast to LONG is the problem, I think.

steffengy commented:
I first thought that too, but `pr30778.c` explicitly calls `__builtin_offsetof `

cznic commented:
You're right.

cznic commented:
Got it. The first line of pr30778.c is

`extern void *memset (void *, int, unsigned long);`

And unsigned long is 64 bit on linux_amd64, but 32 bit on windows_amd64.

So the memset in crt must be specialized for windows_amd64.

steffengy commented:
Oh that's kinda ironic that the error is in the testfile itself xD Thanks for looking into it

steffengy commented:
Another issue:

It seems like `HMODULE` is used as `*T7915855`.

Can this type name change (I need to genereate a function which takes that as a parameter)?

cznic commented:
Yes, it can change. So I suggest to export THMODULE using the IR type specifier of T7915855 from crt and add it to the list of named types in [ccgo here](https://github.com/cznic/ccgo/blob/fa068333a050052e8a95785392e4a55453760def/ccgo.go#L113).

Then define XHMODULE by copy/pasting the current T7915855 definition to some foo_windows.go file in crt.

steffengy commented:
How do I get `T7915855` from the CRT when running `go test` ?

cznic commented:
You have to get it from ccgo by using -log which will dump the Go code in which you can find the definition.

steffengy commented:
Initial version where CCGO's test (including SQLite) pass is committed.

Handling `size_t` for win32 will be fun, so still quite a bit to do.

cznic commented:
> Initial version where CCGO's test (including SQLite) pass is committed.

Well, done!

> Handling size_t for win32 will be fun, so still quite a bit to do.

Yeah, but IMO you've in just one afternoon managed to get the majority of the work completed ;-)

steffengy commented:
Rethinking this after landing CCGO, I now think this is the wrong approach.

I think a better one would be to simply modify mattn's sqlite to use CCGO-sqlite
bindings in a separate fork, while this remains a "simpler" implementation for
testing.

cznic commented:
> I think a better one would be to simply modify mattn's sqlite to use CCGO-sqlite
bindings in a separate fork, while this remains a "simpler" implementation for
testing.

Interesting idea, makes a lot of sense.
